### PR TITLE
Add http check to internet test

### LINKED ISSUE #51145: e2e: Internete connection test should not rely on ICMP achievement
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

When operating conformance test for k8s cluster system on our company internal, only one test "should provide Internet connection for containers [Conformance]" [1] fails, because ICMP is denied between the internet and the company inside network.

[1]: https://github.com/kubernetes/kubernetes/blob/6858bb4e6025218b2f4013173dc027709819bc8c/test/e2e/network/networking.go#L48

**What you expected to happen**:

The test "should provide Internet connection for containers [Conformance]" should succeed on this environment because http/https is reachable from the company internal.
It would be better to use http/https to verify the Internet connection as a general way instead of ICMP.

**How to reproduce it (as minimally and precisely as possible)**:

go run hack/e2e.go -- -v --test --test_args="--ginkgo.focus=\[Conformance\]"

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by oomichi**
**What this PR does / why we need it**:

When operating conformance test for k8s cluster system on a company
internal, only one test "should provide Internet connection for
containers [Conformance]" fails, because ICMP is denied between the
internet and the company inside network. This kind of thing is common
between data centors and the internet also.
So this PR adds a http check to the internet test as a fallback.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:

fixes #51145

**Special notes for your reviewer**:

NOTE: The reasons why using http://google.com are
      - https connection requires another openssl package as [1]
      - the original 8.8.8.8 is the google DNS address

[1]: https://github.com/google/cadvisor/issues/1131

**Release note**:
`NONE`

### CODE REVIEW

**Comment by ncdc** [comment]
Seems somewhat reasonable, although I'm guessing someone else will probably break because they require an http proxy to reach the internet...

cc @timothysc

**Comment by dims** [comment]
@oomichi - do you want to change the logic to try the icmp first and then if that fails then try the url? Also when we use the url we are expecting the dns resolution to work as well which was not there before

**Comment by oomichi** [comment]
@ncdc That is a nice point. I think @dims's idea will cover both cases. Thanks dims!

**Comment by fejta** [comment]
@kubernetes/sig-network-pr-reviews

File: test/e2e/framework/util.go [hunk 1]
```diff
@@ -4474,7 +4462,34 @@ func CheckConnectivityToHost(f *Framework, nodeName, podName, host string, timeo
 			Logf("pod %s/%s logs:\n%s", f.Namespace.Name, pod.Name, logs)
 		}
 	}
+	return err
+}
 
+// CheckConnectivityToHost launches a pod to test connectivity to the specified
+// host. An error will be returned if the host is not reachable from the pod.
+//
+// An empty nodeName will use the schedule to choose where the pod is executed.
```
Comment (caseydavenport) [inline test/e2e/framework/util.go]: typo: schedule -> scheduler

**Review by caseydavenport — COMMENTED** [review]
I think this generally looks good to me.

Slightly concerned about the new requirement for DNS, but I think that's probably OK.

File: test/e2e/framework/util.go (hunk 1 above)
Comment (oomichi) [reply test/e2e/framework/util.go]: Thanks for your review, @caseydavenport 

A nice catch, that was the original typo by moving the comment. This is a good chance to fix it also.

File: test/e2e/network/networking.go [hunk 2]
```diff
@@ -46,9 +46,9 @@ var _ = SIGDescribe("Networking", func() {
 	})
 
 	It("should provide Internet connection for containers [Conformance]", func() {
-		By("Running container which tries to ping 8.8.8.8")
+		By("Running container which tries to reach www.google.com")
 		framework.ExpectNoError(
-			framework.CheckConnectivityToHost(f, "", "ping-test", "8.8.8.8", 30))
+			framework.CheckConnectivityToHost(f, "", "www.google.com", 30))
```
Comment (timothysc) [inline test/e2e/network/networking.go]: Can't do this... 
We need [Conformance] to run on air-gapped environments.

**Review by timothysc — CHANGES_REQUESTED** [review]
You can add a separate test which I actually think there is one for this in the networking tests, but we can't add this restriction.

File: test/e2e/network/networking.go (hunk 2 above)
Comment (oomichi) [reply test/e2e/network/networking.go]: Sorry, I could not follow the meaning of "air-gapped environments".
Could you rephrase ?

Comment (caseydavenport) [reply test/e2e/network/networking.go]: IIUC the test as it is phrased today won't work on air-gapped environments since it tries to hit 8.8.8.8.

Comment (timothysc) [reply test/e2e/network/networking.go]: Several clusters do not allow traffic or resolution to the outside world.  You can make a separate test, but it CAN NOT use the keyword [Conformance]

Comment (oomichi) [reply test/e2e/network/networking.go]: The original 8.8.8.8 is the global address of google DNS, that already is traffic to the outside world from clusters.

Comment (timothysc) [reply test/e2e/network/networking.go]: Hmm I'll have to fix this in a separate patch.

**Review by timothysc — APPROVED** [review]

File: test/e2e/network/networking.go (hunk 2 above)
Comment (caseydavenport) [reply test/e2e/network/networking.go]: I think it's fundamentally a question of whether or not the test in question should be a conformance test.  I'm not sure who defines what Conformance is today.

i.e. the test in question is called `should provide Internet connection for containers [Conformance]`

Comment (timothysc) [reply test/e2e/network/networking.go]: I'll open a separate issue on this.

Comment (oomichi) [reply test/e2e/network/networking.go]: Yeah, I had the same question when I found this issue.
I am not sure it is necessary to ping to the internet as one of conformance tests, because ICMP could be denied in many environments as the commit message.

Oh, sorry I opened the issue for that as https://github.com/kubernetes/kubernetes/issues/52099
If conflict, I will drop mine.

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by spiffxp** [comment]
@kubernetes/sig-architecture-pr-reviews since this touches a [Conformance] test

File: test/e2e/framework/util.go [hunk 3]
```diff
@@ -4491,7 +4479,34 @@ func CheckConnectivityToHost(f *Framework, nodeName, podName, host string, timeo
...
+
+	podName := "ping-test"
+	cmdPing := []string{
+		"ping",
+		"-c", "3", // send 3 pings
+		"-W", "2", // wait at most 2 seconds for a reply
+		"-w", strconv.Itoa(timeout),
+		host,
+	}
+	err := checkConnectivityWithCommand(f, nodeName, podName, cmdPing)
```
Comment (timothysc) [inline test/e2e/framework/util.go]: @jdumars This is the issue that is being tracked.

File: test/e2e/framework/util.go [hunk 4]
```diff
@@ -4491,7 +4479,34 @@ func CheckConnectivityToHost(f *Framework, nodeName, podName, host string, timeo
...
+}
 
+// CheckConnectivityToHost launches a pod to test connectivity to the specified
+// host. An error will be returned if the host is not reachable from the pod.
+//
+// An empty nodeName will use the scheduler to choose where the pod is executed.
+func CheckConnectivityToHost(f *Framework, nodeName, host string, timeout int) error {
+
+	podName := "ping-test"
+	cmdPing := []string{
```
Comment (timothysc) [inline test/e2e/framework/util.go]: @oomichi Should we just remove the ping?

Comment (oomichi) [reply test/e2e/framework/util.go]: @timothysc 
I guess it would work fine in most cases/environments by just removing the ping test.
However I feel it is safe to keep the test to avoid failing in few environments. Maybe I am a little conservative.

File: test/e2e/framework/util.go (hunk 3 above)
Comment (jdumars) [reply test/e2e/framework/util.go]: ack

**Comment by bgrant0607** [comment]
As per #52098, we should remove the conformance label from this test.

**Comment by oomichi** [comment]
https://github.com/kubernetes/kubernetes/pull/53381 has fixed the root problem, so I don't think it is worth to maintain this anymore.

**Comment by jpbetz** [comment]
Removing 1.8 milestone since this has been abandoned.
